domain/src/main/java/org/oppia/android/domain/exploration/ExplorationProgressController.kt

```diff
@@ -814,7 +825,12 @@ class ExplorationProgressController @Inject constructor(
       check(explorationProgress.playStage != SUBMITTING_ANSWER) {
         "Cannot navigate to a next state if an answer submission is pending."
       }
-      explorationProgress.stateDeck.navigateToNextState()
+
+      if (explorationProgress.stateDeck.doesCurrentStateNeedToRevisitOldState()) {
```

Comment:
Ah, we can avoid this, I think. We should introduce a new method in the controller to jump to a specific, previous state. The more we can avoid having to keep state deck & controller state in sync, the easier it will be to maintain this code.

We can keep tests on the new method minimal since we plan to remove it in the full version of this project.